src/domain/auth-rebac/domain/src/repos/rebac_repository.rs

```diff
@@ -0,0 +1,438 @@
+    // Properties
+
+    async fn set_entity_property(
+        &self,
+        entity: &Entity,
+        property: &Property,
+    ) -> Result<(), SetEntityPropertyError>;
+
+    async fn delete_entity_property(
+        &self,
+        entity: &Entity,
+        property_name: PropertyName,
+    ) -> Result<(), DeleteEntityPropertyError>;
+
+    async fn get_entity_properties(
+        &self,
+        entity: &Entity,
+    ) -> Result<Vec<Property>, GetEntityPropertiesError>;
+
+    // Relations
+
+    async fn insert_entities_relation(
+        &self,
+        subject_entity: &Entity,
+        relationship: Relation,
+        object_entity: &Entity,
+    ) -> Result<(), InsertEntitiesRelationError>;
+
+    async fn delete_entities_relation(
+        &self,
+        subject_entity: &Entity,
+        relationship: Relation,
+        object_entity: &Entity,
+    ) -> Result<(), DeleteEntitiesRelationError>;
+
+    async fn get_subject_entity_relations(
+        &self,
+        subject_entity: &Entity,
+    ) -> Result<Vec<ObjectEntityWithRelation>, SubjectEntityRelationsError>;
+
+    async fn get_subject_entity_relations_by_object_type(
+        &self,
+        subject_entity: &Entity,
+        object_entity_type: EntityType,
+    ) -> Result<Vec<ObjectEntityWithRelation>, SubjectEntityRelationsByObjectTypeError>;
+
+    async fn get_relations_between_entities(
+        &self,
+        subject_entity: &Entity,
+        object_entity: &Entity,
+    ) -> Result<Vec<Relation>, GetRelationsBetweenEntitiesError>;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Errors
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Error, Debug)]
+pub enum SetEntityPropertyError {
+    #[error(transparent)]
+    Internal(InternalError),
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Error, Debug)]
+pub enum DeleteEntityPropertyError {
+    #[error(transparent)]
+    EntityNotFound(EntityNotFoundError),
+
+    #[error(transparent)]
+    PropertyNotFound(EntityPropertyNotFoundError),
+
+    #[error(transparent)]
+    Internal(InternalError),
+}
+
+impl DeleteEntityPropertyError {
+    pub fn entity_not_found(entity: &Entity) -> Self {
+        Self::EntityNotFound(entity.into())
+    }
+
+    pub fn property_not_found(entity: &Entity, property_name: PropertyName) -> Self {
+        Self::PropertyNotFound(EntityPropertyNotFoundError {
+            entity_type: entity.entity_type,
+            entity_id: entity.entity_id.to_string(),
+            property_name,
+        })
+    }
+}
+
+#[derive(Error, Debug)]
+#[error("Entity not found: entity_type='{entity_type:?}', entity_id='{entity_id}'")]
+pub struct EntityNotFoundError {
+    pub entity_type: EntityType,
+    pub entity_id: String,
```

Comment:
Thanks for the tip and vector! Realized using `Entity::into_owned()` method

PS. Since `Clone` derive also adds the `ToOwned` implementation, it came out cleaner this way

Reply:
Fixed